README.md
# horizontal_center_date_picker

[![pub package](https://img.shields.io/pub/v/horizontal_center_date_picker.svg)](https://pub.dev/packages/horizontal_center_date_picker)

A flutter widget provides a horizontal date picker and always aligns selected date in center of the widget.

## Usage

This shows Widget's full customizations:
```
HorizontalDatePickerWidget({
    required this.startDate,
    required this.endDate,
    required this.selectedDate,
    required this.widgetWidth,
    required this.datePickerController,
    this.dateItemComponentList = const <DateItem>[
      DateItem.Month,
      DateItem.Day,
      DateItem.WeekDay
    ],
    String? locale,
    this.width = 60,
    this.height = 80,
    this.onValueSelected,
    this.normalColor = Colors.white,
    this.selectedColor = Colors.black,
    this.disabledColor = Colors.white,
    this.normalTextColor = Colors.black,
    this.selectedTextColor = Colors.white,
    this.disabledTextColor = const Color(0xFFBDBDBD),
    this.monthFontSize = 12,
    this.dayFontSize = 18,
    this.weekDayFontSize = 12,
  })
```

1. startDate and endDate are required to control the valid date range.
2. selectedDate is to indicate the default selected date.
3. widgetWidth is the whole widget width
4. width and height is the size of each date option item widget.
5. dateItemComponentList is to indicate the date item label to show and the sequence of each label from top to bottom.
6. locale is the available locale for date formatting. It allows null for using the system default one. Information can be found on https://api.flutter.dev/flutter/date_symbol_data_http_request/availableLocalesForDateFormatting.html 
7. onValueSelected is the callback function when user selected a new date.
8. normalColor, selectedColor and disabledColor are the background color of each date option item widget background.
9. normalTextColor, selectedTextColor and disabledTextColor are the background color of each date option item widget font color.
10. monthFontSize, dayFontSize and weekDayFontSize are the fontsize of the date item label.

## Example

![](horizontal_date_picker.gif)

```
import 'package:flutter/material.dart';
import 'package:horizontal_center_date_picker/datepicker_controller.dart';
import 'package:horizontal_center_date_picker/horizontal_date_picker.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: TestPage(),
    );
  }
}

class TestPage extends StatefulWidget {
  @override
  _TestPageState createState() => _TestPageState();
}

class _TestPageState extends State<TestPage> {
  DatePickerController _datePickerController = DatePickerController();

  @override
  Widget build(BuildContext context) {
    var now = DateTime.now();
    DateTime startDate = now.subtract(Duration(days: 14));
    DateTime endDate = now.add(Duration(days: 7));
    print('startDate = $startDate ; endDate = $endDate');
    return Container(
      color: Colors.grey,
      alignment: Alignment.center,
      child: HorizontalDatePickerWidget(
        locale: 'ja_JP',
        startDate: startDate,
        endDate: endDate,
        selectedDate: now,
        widgetWidth: MediaQuery.of(context).size.width,
        datePickerController: _datePickerController,
        onValueSelected: (date) {
          print('selected = ${date.toIso8601String()}');
        },
      ),
    );
  }
}

```

## License

MIT

CHANGELOG.md
## [2.1.1] - 20220615
* Update: IOS support to 9.0
* Fix: WidgetBinding.instance null check

## [2.1.0] - 20220505
* Update: locale support for month and week of day
* Update: example project
* Update: example android project for 2.10.5 flutter version

## [2.0.0] - 20210502
* Update: Null Safety Support

## [1.0.0+2] - 20200509
* Updated pubspec

## [1.0.0+1] - 20200509
* Change name to avoid same package name on pub.dev

## [1.0.0] - 20200507
* Init version



test/horizontal_date_picker_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:horizontal_date_picker/horizontal_date_picker.dart';

void main() {
  // test('adds one to input values', () {
  //   final calculator = Calculator();
  //   expect(calculator.addOne(2), 3);
  //   expect(calculator.addOne(-7), -6);
  //   expect(calculator.addOne(0), 1);
  //   expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  // });
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:horizontal_center_date_picker/datepicker_controller.dart';
import 'package:horizontal_center_date_picker/horizontal_date_picker.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: TestPage(),
    );
  }
}

class TestPage extends StatefulWidget {
  @override
  _TestPageState createState() => _TestPageState();
}

class _TestPageState extends State<TestPage> {
  DatePickerController _datePickerController = DatePickerController();

  @override
  Widget build(BuildContext context) {
    var now = DateTime.now();
    DateTime startDate = now.subtract(Duration(days: 14));
    DateTime endDate = now.add(Duration(days: 7));
    print('startDate = $startDate ; endDate = $endDate');
    return Container(
      color: Colors.grey,
      alignment: Alignment.center,
      child: HorizontalDatePickerWidget(
        locale: 'ja_JP',
        startDate: startDate,
        endDate: endDate,
        selectedDate: now,
        widgetWidth: MediaQuery.of(context).size.width,
        datePickerController: _datePickerController,
        onValueSelected: (date) {
          print('selected = ${date.toIso8601String()}');
        },
      ),
    );
  }
}


lib/date_item_widget.dart
import 'package:flutter/material.dart';
import 'date_item_state.dart';
import 'package:intl/intl.dart';

import 'date_item.dart';

class DateItemWidget extends StatelessWidget {
  ///display [DateTime]
  final DateTime dateTime;

  ///State of the date
  final DateItemState dateItemState;

  final String locale;

  ///padding of the item widget
  final double padding;

  final double width;
  final double height;

  final double? monthFontSize;
  final double? dayFontSize;
  final double? weekDayFontSize;
  final Color? normalColor;
  final Color? selectedColor;
  final Color? disabledColor;
  final Color? normalTextColor;
  final Color? selectedTextColor;
  final Color? disabledTextColor;

  final List<DateItem> dateItemComponentList;

  DateItemWidget({
    required this.dateTime,
    required this.dateItemState,
    required this.width,
    required this.height,
    required this.dateItemComponentList,
    required this.locale,
    this.padding = 0.0,
    this.normalColor,
    this.selectedColor,
    this.disabledColor,
    this.normalTextColor,
    this.selectedTextColor,
    this.disabledTextColor,
    this.monthFontSize,
    this.dayFontSize,
    this.weekDayFontSize,
  }) : assert(dateItemComponentList.isNotEmpty,
            'dateItemComponentList cannot be empty');

  @override
  Widget build(BuildContext context) {
    return Card(
      margin: EdgeInsets.all(0),
      child: Container(
        width: width + padding,
        height: height,
        padding: EdgeInsets.only(left: padding / 2, right: padding / 2),
        color: _getContainerColorByState(dateItemState),
        alignment: Alignment.center,
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children:
              List<Widget>.generate(this.dateItemComponentList.length, (index) {
            switch (this.dateItemComponentList[index]) {
              case DateItem.WeekDay:
                return Text(DateFormat.E(this.locale).format(this.dateTime),
                    style: TextStyle(
                        color: _getTextColorByState(dateItemState),
                        fontSize: this.weekDayFontSize,
                        fontWeight: _getTextWeightByState(dateItemState)));
              case DateItem.Day:
                return Text(
                  DateFormat.d().format(this.dateTime),
                  style: TextStyle(
                      color: _getTextColorByState(dateItemState),
                      fontSize: this.dayFontSize,
                      fontWeight: _getTextWeightByState(dateItemState)),
                );
              case DateItem.Month:
                return Text(
                  DateFormat.MMM(this.locale).format(this.dateTime),
                  style: TextStyle(
                      color: _getTextColorByState(dateItemState),
                      fontSize: this.monthFontSize,
                      fontWeight: _getTextWeightByState(dateItemState)),
                );
              default:
                return Container();
            }
          }),
        ),
      ),
    );
  }

  Color? _getContainerColorByState(DateItemState state) {
    switch (state) {
      case DateItemState.ACTIVE:
        return normalColor;
      case DateItemState.SELECTED:
        return selectedColor;
      default:
        return disabledColor;
    }
  }

  Color? _getTextColorByState(DateItemState state) {
    switch (state) {
      case DateItemState.ACTIVE:
        return normalTextColor;
      case DateItemState.SELECTED:
        return selectedTextColor;
      default:
        return disabledTextColor;
    }
  }

  FontWeight _getTextWeightByState(DateItemState state) {
    switch (state) {
      case DateItemState.ACTIVE:
        return FontWeight.bold;
      case DateItemState.SELECTED:
        return FontWeight.bold;
      default:
        return FontWeight.normal;
    }
  }
}


lib/datepicker_controller.dart
import 'package:flutter/material.dart';

///controller for scrolling the [HorizontalDatePickerWidget]
class DatePickerController {
  ///scroll to current selected date's position
  ///[isEnableAnimation] default set as true, jump with animation
  void scrollToSelectedItem([bool isEnableAnimation = true]) {
    if (realStartDate != null && selectedDate != null) {
      int index = selectedDate!.difference(realStartDate!).inDays;
      _scrollToSpecificDateByIndex(index);
    }
  }

  ///scroll to specific date's position
  ///[isEnableAnimation] default set as true, jump with animation
  void scrollTo(DateTime dateTime, [bool isEnableAnimation = true]) {
    if (isWithinRange(dateTime) && realStartDate != null) {
      int index = dateTime.difference(realStartDate!).inDays;
      _scrollToSpecificDateByIndex(index);
    } else {
      //out of range do nothing
    }
  }

  ///check date within the start and end date range
  bool isWithinRange(DateTime dateTime) {
    if (startDate != null && endDate != null) {
      return dateTime.compareTo(startDate!) >= 0 &&
          dateTime.compareTo(endDate!) <= 0;
    } else {
      return false;
    }
  }

  ///FIELD: basically for internal use,
  ///if no need to customize as own picker, the following should have no need to modify
  ScrollController? scrollController;
  double shift = 0;

  ///padding + width of Item
  double itemWidth = 0;

  ///the real starting date = the appended header disabled date (index=0)
  DateTime? realStartDate;
  DateTime? selectedDate;
  DateTime? endDate;
  DateTime? startDate;

  ///[index]  listview index
  ///[isEnableAnimation] default set as true, jump with animation
  void _scrollToSpecificDateByIndex(int index,
      [bool isEnableAnimation = true]) {
    var diff = index - shift;
    if (isEnableAnimation) {
      scrollController?.animateTo(diff * itemWidth,
          duration: const Duration(milliseconds: 300), curve: Curves.linear);
    } else {
      scrollController?.jumpTo(diff * itemWidth);
    }
  }
}


lib/horizontal_date_picker.dart
library horizontal_center_date_picker;

import 'package:flutter/material.dart';
import 'package:intl/intl.dart';
import 'package:intl/date_symbol_data_local.dart';
import 'date_item_state.dart';
import 'date_item_widget.dart';
import 'datepicker_controller.dart';

import 'date_item.dart';

class HorizontalDatePickerWidget extends StatefulWidget {
  ///picker start date
  final DateTime startDate;

  ///picker end date
  final DateTime endDate;

  ///default selected date
  final DateTime selectedDate;

  final String locale;

  ///each date item width
  ///if the width is not able to fill the widget's width,
  ///padding will auto append to each item evenly
  final double width;

  ///whole widget's width
  final double widgetWidth;

  ///whole widget's height
  final double height;

  ///callback when a new date selected
  final void Function(DateTime value)? onValueSelected;

  ///controller controls the visible position of the picker
  ///this controller will share both internal and external use
  ///this is required
  final DatePickerController datePickerController;

  ///Date item color and fontsize
  ///fontsize of the month label
  final double monthFontSize;

  ///fontsize of the day label
  final double dayFontSize;

  ///fontsize of the day of week label
  final double weekDayFontSize;

  ///background color of a date item
  final Color normalColor;

  ///background color of a selected date item
  final Color selectedColor;

  ///background color of a disabled date(date that out of the range) item
  final Color disabledColor;

  ///text color of a date item
  final Color normalTextColor;

  ///text color of a selected date item
  final Color selectedTextColor;

  ///text color of a disabled date(date that out of the range) item
  final Color disabledTextColor;

  ///Date item display setting
  ///default set as month, day, day of week, from top to bottom
  ///at least one info must be in the list
  final List<DateItem> dateItemComponentList;

  /// Main widget part of this library.
  /// It is a horizontal date picker that always make the selected option to center.
  HorizontalDatePickerWidget({
    required this.startDate,
    required this.endDate,
    required this.selectedDate,
    required this.widgetWidth,
    required this.datePickerController,
    this.dateItemComponentList = const <DateItem>[
      DateItem.Month,
      DateItem.Day,
      DateItem.WeekDay
    ],

    /// if null, the locale will use the system default one
    /// locale String like "de", can be found via
    /// https://api.flutter.dev/flutter/date_symbol_data_http_request/availableLocalesForDateFormatting.html
    /// or use locale.toString()
    String? locale,
    this.width = 60,
    this.height = 80,
    this.onValueSelected,
    this.normalColor = Colors.white,
    this.selectedColor = Colors.black,
    this.disabledColor = Colors.white,
    this.normalTextColor = Colors.black,
    this.selectedTextColor = Colors.white,
    this.disabledTextColor = const Color(0xFFBDBDBD),
    this.monthFontSize = 12,
    this.dayFontSize = 18,
    this.weekDayFontSize = 12,
  })  : assert(dateItemComponentList.isNotEmpty,
            'dateItemComponentList  cannot be empty'),
        this.locale = locale ?? Intl.systemLocale;

  @override
  _HorizontalDatePickerWidgetState createState() =>
      _HorizontalDatePickerWidgetState(
          this.datePickerController,
          this.widgetWidth,
          this.width,
          this.startDate,
          this.endDate,
          this.selectedDate);
}

class _HorizontalDatePickerWidgetState
    extends State<HorizontalDatePickerWidget> {
  int _itemCount = 0;
  double _padding = 0.0;
  ScrollController _scrollController = ScrollController();

  _HorizontalDatePickerWidgetState(
      DatePickerController controller,
      double ttlWidth,
      double width,
      DateTime startDate,
      DateTime endDate,
      DateTime selectedDate) {
    _init(controller, ttlWidth, width, startDate, endDate, selectedDate);
  }

  @override
  void initState() {
    super.initState();
    initializeDateFormatting(widget.locale, null);
    WidgetsBinding.instance.addPostFrameCallback((_) {
      widget.datePickerController.scrollToSelectedItem();
    });
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      height: widget.height,
      width: widget.widgetWidth,
      child: ListView.builder(
        scrollDirection: Axis.horizontal,
        itemCount: _itemCount,
        controller: _scrollController,
        itemBuilder: (context, index) {
          var dateTime = widget.datePickerController.realStartDate
                  ?.add(Duration(days: index)) ??
              widget.startDate;
          DateItemState dateItemState = _getDateTimeState(dateTime);

          return GestureDetector(
            onTap: () {
              if (dateItemState != DateItemState.DISABLED) {
                widget.datePickerController.selectedDate = dateTime;
                if (widget.onValueSelected != null) {
                  widget.onValueSelected!(dateTime);
                }
                setState(() {
                  widget.datePickerController.scrollToSelectedItem();
                });
              }
            },
            child: DateItemWidget(
              locale: widget.locale,
              dateTime: dateTime,
              padding: _padding,
              width: widget.width,
              height: widget.height,
              dateItemState: dateItemState,
              dayFontSize: widget.dayFontSize,
              monthFontSize: widget.monthFontSize,
              weekDayFontSize: widget.weekDayFontSize,
              normalColor: widget.normalColor,
              selectedColor: widget.selectedColor,
              disabledColor: widget.disabledColor,
              normalTextColor: widget.normalTextColor,
              selectedTextColor: widget.selectedTextColor,
              disabledTextColor: widget.disabledTextColor,
              dateItemComponentList: widget.dateItemComponentList,
            ),
          );
        },
      ),
    );
  }

  DateItemState _getDateTimeState(DateTime dateTime) {
    if (_isSelectedDate(dateTime)) {
      return DateItemState.SELECTED;
    } else {
      if (_isWithinRange(dateTime)) {
        return DateItemState.ACTIVE;
      } else {
        return DateItemState.DISABLED;
      }
    }
  }

  bool _isSelectedDate(DateTime dateTime) {
    return dateTime.year == widget.datePickerController.selectedDate?.year &&
        dateTime.month == widget.datePickerController.selectedDate?.month &&
        dateTime.day == widget.datePickerController.selectedDate?.day;
  }

  void _init(DatePickerController controller, double ttlWidth, double width,
      DateTime startDate, DateTime endDate, DateTime selectedDate) {
    int maxRowChild = 0;
    int shift = 0;
    double shiftPos;
    double widgetWidth = ttlWidth;

    maxRowChild = (widgetWidth / width).floor();

    if (maxRowChild.isOdd) {
      shift = ((maxRowChild - 1) / 2).floor();
      shiftPos = shift.toDouble();
    } else {
      shift = (maxRowChild / 2).floor();
      shiftPos = shift - 0.5;
    }

    //calc padding(L+R)
    _padding = (widgetWidth - (maxRowChild * width)) / maxRowChild;

    _itemCount = shift * 2 + endDate.difference(startDate).inDays + 1;

    controller.scrollController = _scrollController;
    controller.shift = shiftPos;
    controller.itemWidth = _padding + width;

    controller.realStartDate = startDate.subtract(Duration(days: shift));
    controller.selectedDate = selectedDate;
    controller.endDate = endDate;
    controller.startDate = startDate;
  }

  bool _isWithinRange(DateTime dateTime) {
    return widget.datePickerController.isWithinRange(dateTime);
  }
}
